File: 0000001_twoSum/go/solution.go
package twosum

// TwoSum returns indices of two numbers whose sum equals the target.
// Uses a map to store visited numbers and find complements in O(1),
// resulting in an overall O(n) time complexity and O(n) space complexity.
func TwoSum(nums []int, target int) []int {
	visited := make(map[int]int, len(nums))

	for i, n := range nums {
		if idx, exist := visited[target-n]; exist {
			return []int{idx, i}
		}

		visited[n] = i
	}

	return []int{}
}

File: 0000001_twoSum/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

#include "solution.hpp"

using namespace std;

// twoSum returns indices of two numbers whose sum equals the target.
// Uses a hash map to store visited numbers and find complements in O(1),
// resulting in an overall O(n) time complexity and O(n) space complexity.
std::vector<int> Solution::twoSum(std::vector<int>& nums, int target) {
    std::unordered_map<int, int> m;

    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int complement = target - nums[i];

        if (m.find(complement) != m.end()) {
            return {m[complement], i};
        }

        m[nums[i]] = i;
    }

    return {};
}